types/webextension-polyfill/index.d.ts

```diff
@@ -0,0 +1,187 @@
+export { PrivacyNetwork } from "./namespaces/privacy_network";
+export { PrivacyServices } from "./namespaces/privacy_services";
+export { PrivacyWebsites } from "./namespaces/privacy_websites";
+export { Proxy } from "./namespaces/proxy";
+export { Runtime } from "./namespaces/runtime";
+export { Search } from "./namespaces/search";
+export { Sessions } from "./namespaces/sessions";
+export { SidebarAction } from "./namespaces/sidebarAction";
+export { Storage } from "./namespaces/storage";
+export { Tabs } from "./namespaces/tabs";
+export { Theme } from "./namespaces/theme";
+export { TopSites } from "./namespaces/topSites";
+export { Types } from "./namespaces/types";
+export { Urlbar } from "./namespaces/urlbar";
+export { UserScripts } from "./namespaces/userScripts";
+export { WebNavigation } from "./namespaces/webNavigation";
+export { WebRequest } from "./namespaces/webRequest";
+export { Windows } from "./namespaces/windows";
+export { Scripting } from "./namespaces/scripting";
+
+export interface Browser {
+    activityLog: ActivityLog.Static;
+    alarms: Alarms.Static;
+    bookmarks: Bookmarks.Static;
+    action: Action.Static;
+    browserAction: BrowserAction.Static;
+    browserSettings: BrowserSettings.Static;
+    browsingData: BrowsingData.Static;
+    captivePortal: CaptivePortal.Static;
+    clipboard: Clipboard.Static;
+    commands: Commands.Static;
+    contentScripts: ContentScripts.Static;
+    contextualIdentities: ContextualIdentities.Static;
+    cookies: Cookies.Static;
+    declarativeContent: DeclarativeContent.Static;
+    devtools: Devtools.Static;
+    dns: Dns.Static;
+    downloads: Downloads.Static;
+    events: Events.Static;
+    experiments: Experiments.Static;
+    extension: Extension.Static;
+    extensionTypes: ExtensionTypes.Static;
+    find: Find.Static;
+    geckoProfiler: GeckoProfiler.Static;
+    history: History.Static;
+    i18n: I18n.Static;
+    identity: Identity.Static;
+    idle: Idle.Static;
+    management: Management.Static;
+    manifest: Manifest.Static;
+    contextMenus: ContextMenus.Static;
+    menus: Menus.Static;
+    networkStatus: NetworkStatus.Static;
+    normandyAddonStudy: NormandyAddonStudy.Static;
+    notifications: Notifications.Static;
+    omnibox: Omnibox.Static;
+    pageAction: PageAction.Static;
+    permissions: Permissions.Static;
+    pkcs11: Pkcs11.Static;
+    privacy: Privacy.Static;
+    proxy: Proxy.Static;
+    runtime: Runtime.Static;
+    search: Search.Static;
+    sessions: Sessions.Static;
+    sidebarAction: SidebarAction.Static;
+    storage: Storage.Static;
+    tabs: Tabs.Static;
+    theme: Theme.Static;
+    topSites: TopSites.Static;
+    types: Types.Static;
+    urlbar: Urlbar.Static;
+    userScripts: UserScripts.Static;
+    webNavigation: WebNavigation.Static;
+    webRequest: WebRequest.Static;
+    windows: Windows.Static;
+    scripting: Scripting.Static;
+}
+
+declare const browser: Browser;
+
+export default browser;
```

Comment:
webextension-polyfill is not an ES module; there is no default export; it’s a UMD module that assigns to `module.exports` if available. This should be an `export =`, and there should also be an `export as namespace browser;` to make it available as a global. You’ll have to restructure your type exports to be bundled up into a `browser` namespace to do this correctly.

Reply:
Thanks for the review @andrewbranch!

Can you give me a little help on this? I've tried the following approaches without success:

```typescript
declare const browser: Browser;

export = browser; // Error: An export assignment cannot be used in a module with other exported elements
export as namespace browser;
```

Removing all other exports aside from browser does remove the error, but it also removes the `Browser` interface from the exports **and** all the namespaces imported from the namespaces folder, which I still want.

I then tried this approach:
```typescript
declare const browser: WebExtensionPolyFill.Browser;

export = browser;

export as namespace browser;

declare namespace WebExtensionPolyFill {
    export import ActivityLog = ActivityLog_;
// ...

    interface Browser {
// ...
    }
}
```

But got an error importing namespace ActivityLog: `
```typescript
import browser, { ActivityLog } from "webextension-polyfill"; // Error: ... has no exported member 'ActivityLog'
```

Finally, I tried this one, which almost works:
```typescript
declare namespace Browser {
    export import ActivityLog = ActivityLog_;
// ...

    const activityLog: ActivityLog.Static;
// ...
}

export = Browser;
export as namespace browser;
```

This does work for the "~default" import of `browser` and all namespaces, but this way I don't have `Browser` as importable interface anymore.

I have one solution that does work, but it's a bit ugly, since it essentially duplicates every property in the Browser interface (once for the `Browser` Namespace and once for the `Browser` interface):
```typescript

declare namespace Browser {
    export import ActivityLog = ActivityLog_;
// ...

    const activityLog: ActivityLog.Static;
// ...

    export interface Browser {
        activityLog: ActivityLog.Static;
// ...
    }

}

export = Browser;
export as namespace browser;

```

Reply:
I've updated the PR to the last approach, since it works at least and I generate the code automatically anyhow.

Reply:
Great, this is in conflict with the tests:

```typescript
import browser, { Tabs } from 'webextension-polyfill'; // can only be default-imported using the 'esModuleInterop' flag
```

```typescript
import * as browser from 'webextension-polyfill';
import type { Tabs } from 'webextension-polyfill';
// no-duplicate-imports             Multiple imports from 'webextension-polyfill' can be combined into one.
```

Am I missing some form where I can combine the two imports?

Reply:
You got all the right pieces, but split up over separate attempts 😄 

Your closest attempt was this (I’ve corrected it in the diff):

> I then tried this approach:
>
> ```diff
> -declare const browser: WebExtensionPolyFill.Browser;
> +declare const browser: browser.Browser;
>
> export = browser;
>
> export as namespace browser;
>
> -declare namespace WebExtensionPolyFill {
> +declare namespace browser {
>     export import ActivityLog = ActivityLog_;
> // ...
>
>     interface Browser {
> // ...
>     }
> }
> ```

Reply:
That doesn't work: `Cannot redeclare block-scoped variable 'browser'`

Reply:
In that scenario, your `browser` namespace should include _only types_. So you should be deleting the `export const` that are the duplicated members of the `Browser` interface.

Reply:
I did remove the `export const ...` statements inside of the namespace. I still get that error. It goes away once I remove the `export import ...`, but then I also lose these re-exported namespaces. Any idea how to solve that?

Reply:
This is how it currently looks with your suggestion applied:
```typescript
import { ActivityLog as ImportedActivityLog } from "./namespaces/activityLog";
// ...

declare const browser: browser.Browser; // Cannot redeclare block-scoped variable 'browser'

declare namespace browser {
    interface Browser {
        activityLog: ActivityLog.Static;
// ...
    }
    // no more const here

    export import ActivityLog = ImportedActivityLog;
// ...
}

export = browser;

export as namespace browser;
```

Reply:
Hmm, that is a little strange, but it looks like the combined `export import` syntax prevents TS from realizing that those re-exported namespaces are also just types. It’s a little annoying, but it seems to work if you split the import and export:

```ts
import ActivityLog = ImportedActivityLog;
export { ActivityLog };
```

(Of course, you can combine all the exports into that single export declaration.)

Reply:
Indeed, that seems to work. But that solution doesn't seem to work with typescript 3.6:

```
ERROR: 3:19   expect  Compile error in typescript@3.6 but not in typescript@3.7.
Fix with a comment '// Minimum TypeScript Version: 3.7' just under the header.
Module '"."' has no exported member 'Tabs'.
```

Also, the code doesn't look much better if you ask me. If it's okay with you, I'd stay with the currently pushed approach, since it has support for older ts versions.

Reply:
Sure, I have no problem with that. You can also use

```diff
- const activityLog: ActivityLog.Static;
+ const activityLog: Browser['activityLog'];
```

if you prefer to ensure the types are in sync.

Reply:
Thanks. As said, they are auto-generated, so there is little room for error.